File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
public class Solution {
    // public String longestPalindrome(String s) {
    //     string ans = "";
    //     int resLen = 0;
    //     for(int i = 0; i < s.length(); i++){
    //         int left = i;
    //         int right = i;

    //         while(left >= 0 && right < s.length() && s[left] == s[right]){
    //             if(right - left + 1 > resLen){
    //                 resLen = right - left + 1;
    //                 ans = s.substr()
    //             }
    //             left--;
    //             right++;
    //         }
    //     }
    // }

    public String longestPalindrome(String s) {
        if (s.length() <= 1) {
            return s;
        }

        int maxLen = 1;
        String maxStr = s.substring(0, 1);

        for (int i = 0; i < s.length(); i++) {
            for (int j = i + maxLen; j <= s.length(); j++) {
                if (j - i > maxLen && isPalindrome(s.substring(i, j))) {
                    maxLen = j - i;
                    maxStr = s.substring(i, j);
                }
            }
        }

        return maxStr;
    }

    private boolean isPalindrome(String str) {
        int left = 0;
        int right = str.length() - 1;

        while (left < right) {
            if (str.charAt(left) != str.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }

        return true;
    }
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        // Step 1: Preprocess the string
        string t = "^#";
        for (char c : s) {
            t += c;
            t += '#';
        }
        t += '$'; // end sentinel

        int n = t.size();
        vector<int> P(n, 0);
        int center = 0, right = 0;

        // Step 2: Manacher's main loop
        for (int i = 1; i < n - 1; ++i) {
            int mirror = 2 * center - i;

            if (i < right)
                P[i] = min(right - i, P[mirror]);

            // Expand around center i
            while (t[i + (1 + P[i])] == t[i - (1 + P[i])])
                ++P[i];

            // Update center and right boundary
            if (i + P[i] > right) {
                center = i;
                right = i + P[i];
            }
        }

        // Step 3: Find the max length and its center
        int maxLen = 0, centerIndex = 0;
        for (int i = 1; i < n - 1; ++i) {
            if (P[i] > maxLen) {
                maxLen = P[i];
                centerIndex = i;
            }
        }

        // Step 4: Extract the result from original string
        int start = (centerIndex - maxLen) / 2; // remove '#' and '^'
        return s.substr(start, maxLen);
    }

};